Source/WebCore/Modules/gamepad/GamepadHapticActuator.cpp

```diff
@@ -91,33 +102,68 @@ void GamepadHapticActuator::playEffect(Document& document, EffectType effectType
     effectParameters.duration = std::min(effectParameters.duration, GamepadEffectParameters::maximumDuration.milliseconds());
 
     m_playingEffectPromise = WTFMove(promise);
-    GamepadProvider::singleton().playEffect(m_gamepad->index(), m_gamepad->id(), effectType, effectParameters, [this, protectedThis = Ref { *this }, document = Ref { document }, playingEffectPromise = m_playingEffectPromise](bool success) {
+    GamepadProvider::singleton().playEffect(m_gamepad->index(), m_gamepad->id(), effectType, effectParameters, [this, protectedThis = makePendingActivity(*this), playingEffectPromise = m_playingEffectPromise](bool success) mutable {
         if (m_playingEffectPromise != playingEffectPromise)
             return; // Was already pre-empted.
-        document->eventLoop().queueTask(TaskSource::Gamepad, [playingEffectPromise = std::exchange(m_playingEffectPromise, nullptr), success] {
+        queueTaskKeepingObjectAlive(*this, TaskSource::Gamepad, [playingEffectPromise = std::exchange(m_playingEffectPromise, nullptr), success] {
             playingEffectPromise->resolve<IDLEnumeration<Result>>(success ? Result::Complete : Result::Preempted);
         });
     });
 }
 
-void GamepadHapticActuator::reset(Document& document, Ref<DeferredPromise>&& promise)
+void GamepadHapticActuator::reset(Ref<DeferredPromise>&& promise)
 {
-    if (!document.isFullyActive() || document.hidden() || !m_gamepad) {
+    auto document = this->document();
+    if (!document || !document->isFullyActive() || document->hidden() || !m_gamepad) {
         promise->resolve<IDLEnumeration<Result>>(Result::Preempted);
         return;
     }
-    if (auto playingEffectPromise = std::exchange(m_playingEffectPromise, nullptr)) {
-        document.eventLoop().queueTask(TaskSource::Gamepad, [playingEffectPromise = WTFMove(playingEffectPromise)] {
-            playingEffectPromise->resolve<IDLEnumeration<Result>>(Result::Preempted);
-        });
-    }
-    GamepadProvider::singleton().stopEffects(m_gamepad->index(), m_gamepad->id(), [document = Ref { document }, promise = WTFMove(promise)]() mutable {
-        document->eventLoop().queueTask(TaskSource::Gamepad, [promise = WTFMove(promise)] {
+    stopEffects([this, protectedThis = makePendingActivity(*this), promise = WTFMove(promise)]() mutable {
+        queueTaskKeepingObjectAlive(*this, TaskSource::Gamepad, [promise = WTFMove(promise)] {
             promise->resolve<IDLEnumeration<Result>>(Result::Complete);
         });
     });
 }
 
+void GamepadHapticActuator::stopEffects(CompletionHandler<void()>&& completionHandler)
+{
+    if (!m_playingEffectPromise)
+        return completionHandler();
+
+    queueTaskKeepingObjectAlive(*this, TaskSource::Gamepad, [playingEffectPromise = std::exchange(m_playingEffectPromise, nullptr)] {
+        playingEffectPromise->resolve<IDLEnumeration<Result>>(Result::Preempted);
```

Comment:
This is not strictly as per spec as it seems the spec asks to set m_playingEffectPromise to undefined in the queued task, not when enqueuing the task.
I think the proposed patch behavior is good though, it might be worth filing a spec issue.

Reply:
Yes, the spec has issues when it comes to this promise, tracked by https://github.com/w3c/gamepad/issues/179.